programs/bpf_loader/src/syscalls/mod.rs

```diff
@@ -2018,12 +2021,32 @@ declare_builtin_function!(
                 invoke_context.get_check_aligned(),
             )?;
             for val in vals.iter() {
-                let bytes = translate_slice::<u8>(
-                    memory_mapping,
-                    val.as_ptr() as u64,
-                    val.len() as u64,
-                    invoke_context.get_check_aligned(),
-                )?;
+                if direct_mapping {
+                    let syscall_context = invoke_context.get_syscall_context()?;
+
+                    let chunk_iter = MemoryChunkIterator::new(memory_mapping, &syscall_context.accounts_metadata, AccessType::Load, val.as_ptr() as u64, val.len() as u64)?;
+
+                    for item in chunk_iter {
+                        let (region, src_addr, chunk_len) = item?;
+
+                        let addr = Result::from(region.vm_to_host(src_addr, chunk_len as u64))?;
```

Comment:
In `fn iter_memory_pair_chunks()` we may want to iterate in reverse, in which case we want to use the last address in the chunk rather than the first.

We could have an `fn iter_memory_chunks()` which does this.